cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultEmitter.java

```diff
@@ -16,54 +16,64 @@
 
 package co.cask.cdap.etl.common;
 
+import co.cask.cdap.api.metrics.Metrics;
 import co.cask.cdap.etl.api.Emitter;
 import co.cask.cdap.etl.api.InvalidEntry;
 import co.cask.cdap.etl.api.StageMetrics;
 
 import java.util.ArrayList;
-import java.util.Collection;
-import java.util.Iterator;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 
 /**
  * Default implementation of {@link Emitter}. Tracks how many records were emitted.
  *
  * @param <T> the type of object to emit
  */
-public class DefaultEmitter<T> implements Emitter<T>, Iterable<T> {
-  private final List<T> entryList;
-  private final List<InvalidEntry<T>> errorList;
-  private final StageMetrics metrics;
+public class DefaultEmitter<T> implements Emitter<T> {
+  private final Map<String, List<T>> entriesMap;
+  private final Map<String, List<InvalidEntry<T>>> errorMap;
+  private final Metrics metrics;
 
-  public DefaultEmitter(StageMetrics metrics) {
-    this.entryList = new ArrayList<>();
-    this.errorList = new ArrayList<>();
+  public DefaultEmitter(Metrics metrics) {
+    this.entriesMap = new HashMap<>();
+    this.errorMap = new HashMap<>();
     this.metrics = metrics;
   }
 
   @Override
-  public void emit(T value) {
-    entryList.add(value);
-    metrics.count("records.out", 1);
+  public void emit(String stageName, T value) {
+    if (!entriesMap.containsKey(stageName)) {
+      entriesMap.put(stageName, new ArrayList<T>());
+    }
+    entriesMap.get(stageName).add(value);
+    metrics.count(stageName + ".records.out", 1);
```

Comment:
previously in our design, we weren't emitting with stageName, and apart from this the context is same for all the entities in pipeline, so realized we can just pass metrics instead of stage metrics.